### PR TITLE
optimize_caching.md: Change project description

### PR DESCRIPTION
**Pull request by palash25**
**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [ ] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [ ] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

File: _projects/optimize_caching.md [hunk 1]
```diff
@@ -62,23 +67,24 @@ increase performance even further.
...
+  file-dict have been replaced with the factory objects.
+* `Directory` class has been implemented.
 * The corresponding code is properly tested, documented, reviewed, and merged.
 
 ##### CODING PHASE 2
 
-* Section-based caching has been implemented.
+* The Proposed change to ignore unchanged directories has been implemented.
+* Thorough tests have been performed to concretely ascertain a measurable  
+  performance gain
```
Comment (jayvdb) [inline _projects/optimize_caching.md]: missing period at end of item

File: _projects/optimize_caching.md [hunk 2]
```diff
@@ -40,10 +43,12 @@ this greatly improved coala's startup speed.
...
 with their work quicker. And it's fun looking for performance bottlenecks!
 
-In this project, you're to improve on the current strategies. Some enhancements
-such as section-level caching, globbing improvements to not expand on
-unmodified directories, and others have been preliminarily explored. Please
-refer to the related issues to learn more about these.
+In this project, you're to come up with improved versions of the current caching
+strategies used by coala's core and integrate these with the NextGen-Core.
+Some enhancements such as ignoring unmodified directories and use of
+`FileFactory` and `Directory` classes have been already explored and are to be
```
Comment (jayvdb) [inline _projects/optimize_caching.md]: These changes depend on the cEP being accepted.

File: _projects/optimize_caching.md [hunk 3]
```diff
@@ -62,23 +67,24 @@ increase performance even further.
...
 * The corresponding code has been thoroughly tested, documented, reviewed, and  
   merged.
-* At least one of the proposed performance bottleneck improvements have been  
-  implemented, tested, documented, and merged.
 
 ##### CODING PHASE 3
 
-* At least an additional proposed performance bottleneck improvement has been  
-  implemented, tested, documented, and merged.
+* All the remaining cache control flags like `--cache-optimize`,
```
Comment (jayvdb) [inline _projects/optimize_caching.md]: depends on cEP being accepted.

Comment (palash25) [reply _projects/optimize_caching.md]: @jayvdb the cEP has been merged https://github.com/coala/cEPs/pull/146/

Can we merge this too?

**Comment by palash25** [comment]
Weird this error wasn't there when I pushed it for the first time.

**Comment by jayvdb** [comment]
It is because that repo was moved.

Needs a new issue and PR to fix it.

File: _projects/optimize_caching.md [hunk 4]
```diff
@@ -1,25 +1,29 @@
...
 difficulty: medium
 initiatives:
   - GSoC
   - RGSoC
 issues:
-  - "https://github.com/coala/coala/issues/2541"
-  - "https://github.com/coala/coala/issues/2699"
+  - "https://github.com/coala/coala/issues/3530"
   - "https://github.com/coala/coala/issues/2912"
+  - "https://github.com/coala/coala/issues/2541"
```
Comment (jayvdb) [inline _projects/optimize_caching.md]: why moved?

File: _projects/optimize_caching.md [hunk 5]
```diff
@@ -1,25 +1,29 @@
...
   - shreyans800755
   - fneu
 name: "Optimize Caching"
 requirements:
   - "At least one patch to the coala core should be accepted and merged."
-  - "The student should be familiar with ongoing technical cEPs especially [cEP-0002](https://github.com/coala/cEPs/blob/master/cEP-0002.md)."
-  - "The proposal needs to show some familiarity with coala's caching implementation."
+  - "The student should be familiar with ongoing technical cEPs especially
+    [cEP-0002](https://github.com/coala/cEPs/blob/master/cEP-0002.md)."
+  - "The proposal needs to show some familiarity with coala's caching
```
Comment (jayvdb) [inline _projects/optimize_caching.md]: why these changes?

File: _projects/optimize_caching.md [hunk 6]
```diff
@@ -52,8 +59,8 @@ increase performance even further.
 
 ##### PREPARATION/BONDING
 
-* The applicant has identified at least 3 other performance bottlenecks that can be  
-  improved. These can be smaller improvements, focussing on one particular  
+* The applicant has identified at least 3 other performance bottlenecks that can
+  be improved. These can be smaller improvements, focussing on one particular
```
Comment (jayvdb) [inline _projects/optimize_caching.md]: revert?

File: _projects/optimize_caching.md [hunk 7]
```diff
@@ -62,23 +69,24 @@ increase performance even further.
 
 ##### CODING PHASE 1
 
-* The Proposed change to not explore unchanged directories has been
```
Comment (jayvdb) [inline _projects/optimize_caching.md]: why is this removed?  Is the `Directory` class related to this?  if so, it would be better to add `Directory` to this first bullet point, revising it a bit, so it is obvious you are clarifying/revising rather than removing a deliverable.

File: _projects/optimize_caching.md [hunk 8]
```diff
@@ -62,23 +69,24 @@ increase performance even further.
 
 ##### CODING PHASE 1
 
-* The Proposed change to not explore unchanged directories has been  
-  implemented.
-* Thorough tests have been performed to concretely ascertain a measurable
```
Comment (jayvdb) [inline _projects/optimize_caching.md]: I'm not keen on this being moved out of the project coding phase 1, especially not out of the project brief.

moving it now is planning to fail.

You should be trying to achieve this.

If you fall short, the mentors might push this task into coding phase 2.

File: _projects/optimize_caching.md [hunk 9]
```diff
@@ -1,25 +1,29 @@
 ---
 collaborating_projects:
   - coala
-desc: "Improve coala's performance by improving file caching and other performance bottlenecks."
+desc: "Improve coala's performance by improving file loading using factory
```
Comment (jayvdb) [inline _projects/optimize_caching.md]: IMO, aim to make it shorter, not longer.

File: _projects/optimize_caching.md (hunk 8 above)
Comment (palash25) [reply _projects/optimize_caching.md]: I have already discussed this with @Makman2 ignore directories functionality will be implemented during phase 2 also we are still not clear on a few implementation details so we had decided to take this up in phase2.

Also the same timeline was mentioned in my GSoC proposal and the cEP which got merged so I think we should keep the timelines consistent across all proposals

Comment (palash25) [reply _projects/optimize_caching.md]: Adding some further clarification as to why it was moved out of phase 1:
The original description included section based caching for phase2 since that was ruled during discussion before GSoC see: https://github.com/coala/coala/issues/2699#issuecomment-374154490 and the ignore directories depending upon factory class implementation it was moved to phase 2 and replaced by factory classes implementation in phase 1 which has been done and waiting to be approved See: https://github.com/coala/coala/pull/5485  & https://github.com/palash25/coala/pull/6

File: _projects/optimize_caching.md (hunk 4 above)
Comment (palash25) [reply _projects/optimize_caching.md]: Because I wanted to keep the sequence of issues in line with the timeline
The first is about `FileFactory` (implemented)
The second is  a general thread about Performance Optimization (one of them being implementing a new file caching module to replace linecache being used in the current implementation of `FileFactory`; to be implement in phase2 week1)
The third (which was moved) ignore directory will take the majority of the coding phase 2
The last one is cache control (phase3)

File: _projects/optimize_caching.md (hunk 5 above)
Comment (palash25) [reply _projects/optimize_caching.md]: my editor did that. 81 char limit.

File: _projects/optimize_caching.md (hunk 7 above)
Comment (palash25) [reply _projects/optimize_caching.md]: Done. Made Directory the first bullet point and added extra explaination.

File: _projects/optimize_caching.md (hunk 8 above)
Comment (jayvdb) [reply _projects/optimize_caching.md]: This item was about measuring performance at beginning of project , so comparisons can be made after changes are made.
This is mandatory for verifying the project is successful in each phase

Comment (palash25) [reply _projects/optimize_caching.md]: I'm sorry I misunderstood. I thought you were talking about the first point being removed.

File: _projects/optimize_caching.md [hunk 10]
```diff
@@ -62,23 +66,26 @@ increase performance even further.
 
 ##### CODING PHASE 1
 
-* The Proposed change to not explore unchanged directories has been  
-  implemented.
-* Thorough tests have been performed to concretely ascertain a measurable  
+* `Directory` class has been implemented.
+* `FileFactory` class has been implemented and the file contents in the
+  file-dict have been replaced with the factory objects. (Both these classes would later be used for implementing ignore directories functionality)
+* Thorough tests have been performed to concretely ascertain a measurable
```
Comment (jayvdb) [inline _projects/optimize_caching.md]: Why is this line changing?

File: _projects/optimize_caching.md (hunk 4 above)
Comment (Makman2) [reply _projects/optimize_caching.md]: Okay with me

File: _projects/optimize_caching.md (hunk 2 above)
Comment (Makman2) [reply _projects/optimize_caching.md]: @palash25 you can link to the old version of cEP-0002 that still contains the `FileProxy`, then you avoid describing features ahead of this cEP.

File: _projects/optimize_caching.md (hunk 3 above)
Comment (Makman2) [reply _projects/optimize_caching.md]: Even without cEP acceptance, those cache-control flags are needed in any way.

Comment (Makman2) [reply _projects/optimize_caching.md]: Maybe not specific ones like `--cache-optimize`, but `--cache-compression` for example is for nearly every cache implementation viable.

File: _projects/optimize_caching.md [hunk 11]
```diff
@@ -62,23 +66,26 @@ increase performance even further.
...
 ##### CODING PHASE 1
 
-* The Proposed change to not explore unchanged directories has been  
-  implemented.
-* Thorough tests have been performed to concretely ascertain a measurable  
+* Thorough tests have been performed to concretely ascertain a measurable
   performance gain.
+* `Directory` class has been implemented.
+* `FileFactory` class has been implemented and the file contents in the
+  file-dict have been replaced with the factory objects. (Both these classes would later be used for implementing ignore directories functionality)
```
Comment (jayvdb) [inline _projects/optimize_caching.md]: line length.

**Comment by jayvdb** [comment]
needs rebase.

**Comment by jayvdb** [comment]
ack https://github.com/coala/projects/pull/578/commits/a228c2b39ef17b3c6d3572853601dfa05b938173 https://github.com/coala/projects/pull/578/commits/759f74f88206c58ac80dd7acd12cadbae13f28cb

**Comment by jayvdb** [comment]
@gitmate-bot ff
